Make our types definition approach more consistent

pablopareja opened the pull request:
This PR includes the following updates/additions:

- `types` top-level folder
- consistent definition of types all throughout the repository by means of `./types.ts` files in each component folder _(mainly to store the component-specific prop types definitions)_ 
- consistent definition of constants all throughout the repository by means of `./constants.ts` files in each component folder _(only when necessary)_

davidsingal commented:
Awesome! Please, could you rename the files `types.ts` to `types.d.ts`? It will be even more consistent.

davidsingal commented on components/button/component.tsx:
```diff
@@ -1,58 +1,10 @@
-import { ButtonHTMLAttributes, AnchorHTMLAttributes, FC } from 'react';
-import Link, { LinkProps } from 'next/link';
+import { FC } from 'react';
```
If you use `React.FC` you don't have to declare on your imports.

davidsingal commented on components/avatar/component.tsx:
```diff
@@ -1,11 +1,7 @@
-  className?: string;
-  bgImage?: string;
-}
+import { AvatarProps } from './types';
```
Please consider to use this notation for types: `import type { AvatarProps } from './types';`

pablopareja replied:
Done! ✅

pablopareja commented:
> 
> 
> Awesome! Please, could you rename the files `types.ts` to `types.d.ts`? It will be even more consistent.

Done! 😉

davidsingal commented on components/avatar/types.d.ts:
```diff
@@ -0,0 +1,7 @@
+import { ButtonHTMLAttributes } from 'react';
+
+export interface AvatarProps extends ButtonHTMLAttributes<HTMLButtonElement> {
```
Consider the next piece of code instead:

```
import type { ImgHTMLAttributes, FC } from 'react';
export type AvatarProps = ImgHTMLAttributes<HTMLImageElement> & {
  bgImage?: string;
};
```
The props `children` and `className` are already included in the imgHTMLAttributes type.

davidsingal commented on components/forms/radio/types.d.ts:
```diff
@@ -0,0 +1,7 @@
+import { InputHTMLAttributes } from 'react';
+
+export interface RadioProps extends InputHTMLAttributes<HTMLInputElement> {
```
You can use `type` or `interface` but as a convention, I would recommend using only `interface` for class. For literal objects, arrays, and primitive elements better to use `type`.

pablopareja replied:
@davidsingal Let's talk about this next week if you want 👍 
I don't have a strong opinion regarding `interface` vs `type` but I was having a look at [an article](https://blog.logrocket.com/types-vs-interfaces-in-typescript/) now where they explicitly encouraged the use of interfaces to model component props 
![image](https://user-images.githubusercontent.com/545342/122594325-2c65a580-d067-11eb-9fa5-b1d74c765eab.png)
Happy to hear contrasting opinions though! 😃

davidsingal replied:
I do really like it when someone wants to discuss :). 
I agree on I'ts very opinionated. Also in this article says: 
"Interface work better with objects and method objects, and types are better to work with functions, complex types, etc."
"Interfaces in TypeScript can extend classes, this is a very awesome concept that helps a lot in a more object-oriented way of programming. We can also create classes implementing interfaces."
and
"Interface work better with objects and method objects, and types are better to work with functions, complex types, etc."
We use React as a function instead of a class.
To be honest this is not clear. I'll try to find a better explanation about this to have a stronger opinion of this.

pablopareja replied:
@davidsingal Cool, let me know when you want to talk about it 😉

davidsingal commented on components/map/legend/types/basic/types.d.ts:
```diff
@@ -0,0 +1,7 @@
+export interface LegendTypeBasicProps {
+  className?: string;
+  items: Array<{
```
Consider to create a type only for this:

```
const type LegendTypeItem = {
  value: string;    
  color: string;
};
```
and replace Array to `items: LegendTypeItem[]`

pablopareja replied:
👍

davidsingal commented on components/map/legend/types/gradient/types.d.ts:
```diff
@@ -0,0 +1,7 @@
+export interface LegendTypeGradientProps {
+  className?: string;
+  items: Array<{
```
Related to my comment above, here you can reuse the `LegendTypeItem`

davidsingal commented on components/toast/component.tsx:
```diff
@@ -1,41 +1,14 @@
 import { FC, useCallback, useEffect, useRef } from 'react';
```
Other case where you can replace to:

```
import { useCallback, useEffect, useRef } from 'react';
import type { FC } from 'react';
```
or better, using only: 

`import { useCallback, useEffect, useRef } from 'react';`

and replace `export const Toast: React.FC<ToastProps>`. React.FC is injected by @types/react

davidsingal commented on types/index.ts:
```diff
@@ -0,0 +1,8 @@
+/**
```
Remove this file.

pablopareja replied:
@davidsingal Don't you think this could be useful for developers to understand that "general" types should be defined under this folder?

davidsingal replied:
Do you have an example of case of use? Any case, it should be index.d.ts.

davidsingal commented:
We are still on `strict: false` mode. In case you need more fun try to enable it :). Anyway, really good job!

mbarrenechea reviewed (approved):
Amazing work!

pablopareja commented:
@davidsingal Hey! 👋 
I've addressed most of the comments that you left as part of your last review _(thank you for that by the way! 👏 )_
I also set `strict: true` and only had to make one change ✊ 
Could you have a look at the changes and let me know if you think it's ready for merging?

PS: we can always review the `interface` vs `type` discussion later and modify the definitions appropriately in a subsequent PR if necessary.

davidsingal commented:
I agree we can merge it and then apply the needed changes after.

davidsingal commented on components/avatar/types.d.ts:
```diff
@@ -0,0 +1,7 @@
+import { ButtonHTMLAttributes } from 'react';
+
+export interface AvatarProps extends ButtonHTMLAttributes<HTMLButtonElement> {
+  children?: React.ReactNode;
```
As you are extending ButtonHTMLAttributes you don't need to override `children` and `className`

davidsingal commented:
@pablopareja you cannot trust only on eslint for types check. Typescript checks the types consistency only on build. So, I rebased your branch with the last changes from main where eslint task includes also the types check. You can run `yarn check-types` and see all the errors.
This was my output: Found  157 errors.
We can live with `strict: false` for now hehe

pablopareja commented:
@davidsingal Typescript errors thrown when using `strict: false` fixed with the latest commit 🙌 
I'm going to merge the PR now 🥳 🎈
